kratos/solving_strategies/convergencecriterias/mixed_generic_criteria.h

```diff
@@ -0,0 +1,441 @@
+    }
+
+    /**
+     * @brief Method to check convergence
+     * This method checks the convergence of the provided norms with the user-defined tolerances
+     * @param rConvergenceNorms Tuple containing the absolute and relative convergence values
+     * @return true Convergence is satisfied
+     * @return false Convergence is not satisfied
+     */
+    bool CheckConvergence(const std::tuple<std::vector<TDataType>, std::vector<TDataType>>& rConvergenceNorms)
+    {
+        int is_converged = 0;
+        const auto& var_ratio = std::get<0>(rConvergenceNorms);
+        const auto& var_abs = std::get<1>(rConvergenceNorms);
+
+        for (int i = 0; i < mVariableSize; i++) {
+            const auto r_var_data = mVariableDataVector[i];
+            const int key_map = mLocalKeyMap[r_var_data->Key()];
+            is_converged += var_ratio[key_map] <= mRatioToleranceVector[key_map] || var_abs[key_map] <= mAbsToleranceVector[key_map];
+        }
+
+        if (is_converged) {
+            KRATOS_INFO_IF("", this->GetEchoLevel() > 0) << "*** CONVERGENCE IS ACHIEVED ***" << std::endl;
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    ///@}
+    ///@name Protected  Access
+    ///@{
+
+
+    ///@}
+    ///@name Protected Inquiry
+    ///@{
+
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+
+    ///@}
+private:
+    ///@name Private Static Member Variables
+    ///@{
+
+
+    ///@}
+    ///@name Private Member Variables
+    ///@{
+
+    const int mVariableSize;
+    const std::vector<VariableData*> mVariableDataVector;
+    const std::vector<TDataType> mRatioToleranceVector;
+    const std::vector<TDataType> mAbsToleranceVector;
+    std::unordered_map<KeyType, KeyType> mLocalKeyMap;
+
+    ///@}
+    ///@name Private Operators
+    ///@{
+
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    /**
+     * @brief Get the Norm Values
+     * This function accumulates the solution and increment norm values in the provided arrays.
+     * Note that these arrays are assumed to be already initialized to zero.
+     * @param rModelPart Reference to the ModelPart containing the fluid problem.
+     * @param rDofSet Reference to the container of the problem's degrees of freedom (stored by the BuilderAndSolver)
+     * @param rDx Vector of results (variations on nodal variables)
+     * @param rDofsCount Array containing the number of DOFs per variable
+     * @param rSolutionNormsVector Array containing the solution norms accumulated values for each variable checked
+     * @param rIncreaseNormsVector Array containing the correction norms accumulated values for each variable checked
+     */
+    virtual void GetNormValues(
+        const ModelPart& rModelPart,
+        const DofsArrayType& rDofSet,
+        const TSystemVectorType& rDx,
+        std::vector<int>& rDofsCount,
+        std::vector<TDataType>& rSolutionNormsVector,
+        std::vector<TDataType>& rIncreaseNormsVector)
```

Comment:
shouldn't you clear those?

Reply:
I don't think so. They are declared and initialized outside this function.